File: binary_search_tree/delete_a_node_in_BST/delete_a_node_in_BST.py
# https://www.youtube.com/watch?v=LFzAoJJt92M
# https://www.youtube.com/watch?v=gcULXE7ViZw

def findMostLeft(root):
    cur = root.right
    while cur.left:
        cur = cur.left
    return cur

def deleteNode(root, key):
    if not root:
        return None
    
    if key < root.val:
        root.left = deleteNode(root.left, key)
    elif key > root.val:
        root.right = deleteNode(root.right, key)
    else:
        if not root.left and not root.right:
            return None
        if not root.left:
            return root.right
        if not root.right:
            return root.left
        if root.right and root.left:
            root.val = findMostLeft(root).val
            root.right = deleteNode(root.right, root.val)
            
    return root
        

File: binary_search_tree/delete_a_node_in_BST/delete_a_node_in_BST.js
// helper function for find most left node on right side of root
const findMostLeft = (root) => {
    let cur = root.right
    while (cur.left) {
        cur = cur.left
    }

    return cur
}

const deleteNode = (root, key) => {
    if (root === null) return root
    // determine key is = > < root.val
    if (key < root.val) {
        // update root.left if there has any change
        root.left = deleteNode(root.left, key)
    } else if (key > root.val) {
        // update root.right if there has any change
        root.right = deleteNode(root.right, key)
    } else {
        if (!root.right && !root.left) {
            return null
        } else if (!root.right) {
            // root only has one left child, replace root with root.left
            return root.left
        } else if (!root.left) {
            // root only has one right child, replace root with root.right
            return root.right
        } else {
            // root has both left and right child, replace root with the most smallest value found on the right side
            root.val = findMostLeft(root).val
            // delete the duplicate node on right side
            root.right = deleteNode(root.right, root.val)
        }
    }
    // if no node is equal to key, return root itself
    return root
}